voxblox/include/voxblox/integrator/tsdf_integrator.h

```diff
@@ -65,6 +69,159 @@ class TsdfIntegratorBase {
+          atomic_idx_(0) {
+      DCHECK_EQ(points_C_.size(), colors_.size());
+    }
+
+    bool getNextPoint(size_t* point_idx, Point* point_G, bool* clearing) {
+      while (atomic_idx_.load() < points_C_.size()) {
+        *point_idx = getMixedIdx(atomic_idx_.fetch_add(1));
+        const Point& point_C = points_C_[*point_idx];
+        *point_G = T_G_C_ * point_C;
+
+        FloatingPoint ray_distance = (*point_G - origin_).norm();
+        if (ray_distance < config_.min_ray_length_m) {
+          continue;
+        } else if (ray_distance > config_.max_ray_length_m) {
+          if (config_.allow_clear) {
+            *clearing = true;
+          } else {
+            continue;
+          }
+        } else {
+          *clearing = false;
+        }
+
+        return true;
+      }
+
+      return false;
+    }
+
+    bool getNextPoint(Point* point_C, Point* point_G, Color* color,
+                      Point* origin, bool* clearing) {
+      size_t point_idx;
+      if (getNextPoint(&point_idx, point_G, clearing)) {
+        *origin = origin_;
+        *point_C = points_C_[point_idx];
+        *color = colors_[point_idx];
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    size_t numPoints() { return points_C_.size(); }
+
+   private:
+    // Mixes up the order rays are given in so that each thread is working with
+    // a
+    // point that is far away from the other threads current points.
+    size_t getMixedIdx(size_t base_idx) {
+      const size_t number_of_points = points_C_.size();
+      const size_t number_of_groups = config_.integrator_threads;
+      const size_t points_per_group = number_of_points / number_of_groups;
+
+      size_t group_num = base_idx % number_of_groups;
+      size_t position_in_group = base_idx / number_of_groups;
+
+      return group_num * points_per_group + position_in_group;
+    }
+
+    const Transformation& T_G_C_;
+    const Pointcloud& points_C_;
+    const Colors& colors_;
+    const Config& config_;
+    const Point origin_;
+    std::atomic<size_t> atomic_idx_;
+  };
+
+  // Can be used to pre-allocate all the blocks needed by the integrator. This
+  // allows thread-safe lockless block map access
+  void allocateBlocks(const Transformation& T_G_C, const Pointcloud& points_C,
+                      const Colors& colors) {
+    IndexSet index_set;
+
+    point_getter_ =
+        std::make_shared<PointGetter>(T_G_C, points_C, colors, config_);
+
+    Point point_C, point_G, origin;
+    Color color;
+    bool clearing;
+    while (point_getter_->getNextPoint(&point_C, &point_G, &color, &origin,
+                                       &clearing)) {
+      // immediately check if the voxel the point is in has already been ray
+      // traced (saves time setting up ray tracer for already used points)
+      AnyIndex block_idx =
+          getGridIndexFromPoint(point_G, layer_->block_size_inv());
+      if (index_set.count(block_idx)) {
```

Comment:
I believe that this combined with the adding neighbors means that all blocks will be allocated. I need to profile to see how significant the speed gain of this is. If it is not significant I will remove it.